Heroic Games Launcher GOG & Epic Support

daemon-mouse opened the pull request:
This patch adds support for installing Mod Organizer 2 for games installed through Heroic Games Launcher. Both the system-installed and flatpak versions of Heroic are supported.

Games from the GOG and Epic stores are enabled. Note that support for Amazon Prime Games is **not** added (as I do not own any games from that store and thus have no data to work with); however, the scaffolding is here for such support to be added with little difficulty later.

The steps of loading game info, cleaning game prefixes, installing MO2 itself, installing a script extender, and the `nxm://` scheme handler have all been given custom support for the new launcher and storefronts. Additionally, support for using winetricks for installing verbs has been added just for Heroic, as we cannot use protontricks as with Steam. For this, we prefer (in descending order of priority) the winetricks downloaded by `download_external_resources.sh`, the winetricks shipped in the `com.github.Matoking.protontricks` flatpak, or the system-installed winetricks. This prioritizes the most likely sources of up-to-date winetricks available to maximize the chances of our verbs succeeding.

It was necessary to make some opinionated changes:

- When setting up a clean wine prefix, the guidance given to the user is to enable the Steam runtime and to use Proton 9.0. In the interest of transparency, it should be noted that this goes against the recommendations of the Heroic developers — namely, that users should use Proton-GE releases on the grounds that this (in fairness, quite excellent) build of Proton contains game-specific protonfixes. However, I believe the advice I've given is better for a few reasons:
  -  It's better general advice: Valve has invested considerable effort making sure games "just work" when a user presses the Play button in Steam, and users are well served by sticking to that configuration unless they have a specific need for Proton-GE.
  - We needn't be so ambitious as to advise that the user set this for all games as a default; we only make this suggestion for the game for which the user is installing MO2.
  - It's the clear choice from the perspective of this application anyway. Using the Steam runtime and Proton 9.0 matches the platform we recommend and support already for Steam games. This advice makes the environment of Heroic games match their Steam counterpart as closely as reasonably possible and therefore minimizes the extra support burden for this project.
- We have replaced the unstructured `appid.txt` file with a more extensible `variables.sh` file for the benefit of the NXM handler. We remain backwards-compatible with instances containing the former. The new file allows us to pass as much information to the NXM handler as we like in a safe and extensible way, which proved necessary for communicating to the handler what launcher the game uses and its product IDs on other storefronts.

This patch adds product IDs for GOG for every game. With the exception of Fallout 4 (which I own and obtained directly from my local install), this information is from [GOG Database](https://www.gogdb.org). Unfortunately, there is no comparable resource for Epic games. Without guesswork, it is only possible to obtain the product ID for an Epic game by purchasing it. We will need to depend on contributions from users who own these games on Epic to fill in values for `game_epic_productid`.

In accordance with the new launcher support, the error cases of a missing `game_appid` and failure to find a Steam installation of a game have been downgraded from an error to a warning, as such a game might be found in another launcher. It should be noted that in this implementation, Steam short-circuits Heroic, in the sense that if a game is installed with Steam, Heroic will not be searched and MO2 will be installed for the Steam game, even if the user has both Steam and Epic/GOG versions installed. We may consider searching every location and presenting a dialog asking the user to choose which version to use with MO2; however, I've left that work out for now as I would consider that a rare case.

Contributes toward the progress of #796.

Furglitch committed:
Update ignored labels for automated build

Furglitch committed:
Add prefix configuration disclaimer

Furglitch committed:
Add verbose logging flag for protontricks

Furglitch committed:
Add datetime to filename

Furglitch committed:
Remove redundant dialog type

Furglitch committed:
Adjust bug report to reflect logging

Furglitch committed:
Prevent redundant output

Furglitch committed:
Fix syntax error

Furglitch committed:
Add verbose output for external downloads

Furglitch committed:
Add checksum validation to MO2 download

Furglitch committed:
Allow multiple attempts for valid checksum

Furglitch committed:
Update installer version to 5.2.8

cobarx committed:
Remove invalid -z evaluator

AceOfKestrels committed:
Create shell.nix

AceOfKestrels committed:
Remove unnecessary flag and dependencies

Actual code does not include pthread.h so this flag is unnecessary

With that we can also get rid of the pthread libraries

AceOfKestrels committed:
Add log files to gitignore

AceOfKestrels committed:
Update shell.nix

daemon-mouse committed:
gamesinfo: add placeholder variables for epic and gog productid

daemon-mouse committed:
fallout4: add GOG product id

Furglitch committed:
Declutter documentation in favor of wiki

Furglitch committed:
Merge branch 'master' into feature/nix-shell

Furglitch committed:
Merge pull request #836 from cobarx/bugfix/z-bash-flag-error

Remove invalid -z evaluator

Furglitch committed:
Merge pull request #839 from AceOfKestrels/feature/nix-shell

Add shell.nix and remove unnecssary build dependency

Furglitch committed:
Fix impossible FO3 downgrade check

Resolves #844

Furglitch committed:
Update version number

Furglitch committed:
Fix syntax error in load_gameinfo.sh

daemon-mouse committed:
fix Fallout 3 downgrade check

This check was formerly unreachable. Further, even if it were hit, it
was checking in the wrong folder for the Fallout3Launcher.exe file.

Rather than relying on the presence or absence of $steam_library, script
find-library-for-file.sh now returns different error codes for the
following cases:

0: Success
1: Game not installed
2: Fallout 3 is installed, but must be downgraded

daemon-mouse committed:
heroic: load game info for GOG and Epic games

This introduces the new helper script find-heroic-game-installation.sh
to find games installed by Heroic. It does a bit more work than its
sister script for Steam, find_library_for_file.sh; namely, it looks up
the Wine version and prefix folder set for the game. We need these
because we can't use protontricks later.

In load_gameinfo.sh, we now search in both Steam and Heroic locations to
find the selected game. The current logic is short-circuiting, so if a
Steam game is found, Heroic will not be checked. We could consider
checking both and allowing the user to choose which version they want;
however, I think the likelihood of someone installing the same game in
two different launchers is low enough to not justify the extra code.

This commit downgrades an absent appid from an error to a warning. This
allows us to incrementally add support for GOG/Epic games before adding
Steam support.

We set a new variable, game_launcher, which has the value of "steam" or
"heroic", allowing other scripts to implement different logic for each
launcher.

daemon-mouse committed:
heroic: fall back to default wine version if unset for a game

Also, fail correctly if wine versions or prefixes are set to the empty
string.

daemon-mouse committed:
clean_game_prefix.sh: add support for heroic game prefixes

daemon-mouse committed:
clean_game_prefix.sh: add Heroic instructions

We advise the user to use Proton 9.0 and the Steam Runtime when running
our game. This resultant configuration matches the conditions under
which we support MO2 for Steam games, minimizing the support burden we
add.

daemon-mouse committed:
heroic: use winetricks to apply verbs

This introduces a new script, winetricks.sh, to wrap the command. Like
protontricks.sh, we prefer to use the version in the
com.github.Matoking.protontricks flatpak if present.

daemon-mouse committed:
print the name of the correct launcher in the last infobox

Furglitch committed:
Merge pull request #845 from daemon-mouse/fo3-downgrade-check-fix

Fix Fallout 3 downgrade check

Furglitch committed:
Update version to 5.2.10

daemon-mouse committed:
select_install_dir.sh: use ~/Games/Heroic as default path with heroic

This is for a nuanced reason: the flatpak version of Heroic can't run
MO2 when it is installed to ~/Games/mod-organizer-2-*. See the added
comment for more info.

daemon-mouse committed:
add heroic support to NXM handler

This makes an important change to our instance directories: we are now
using variables.sh instead of appid.txt.

The latter is inextensible -- we can't just add heroic-related
information to it without using error-prone parsing to determine if it
is a new or old version of that file.

The new file variables.sh can be easily sourced by
modorganizer2-nxm-broker.sh, and can be also extended to add more
information over time as needed.

Some extra code has been added for backwards-compatibility when an
appid.txt file is found.

daemon-mouse committed:
winetricks.sh: use and prefer downloaded winetricks if present

This uses the winetricks copy downloaded in download_external_resources
.sh and installed to ~/.local/share/modorganizer2/winetricks. Curiously,
it seems like this file was completely unused.

I'm happy to put it to use, as the winetricks verbs are more likely to
succeed the newer the version of winetricks we use, and the downloaded
one is the bleeding-edge latest release.

daemon-mouse committed:
GOG: add known product IDs

This commit is best-effort, based on information from
[GOG Database](https://www.gogdb.org). I don't own all these titles and
can't test that they'll be detected correctly.

Future developers: make sure to use the ID matching a search result with
Type = Game, not Package or DLC.

daemon-mouse committed:
script extender: add support for GOG and Epic specific URLs

daemon-mouse committed:
Fallout 4: add F4SE url for GOG release

daemon-mouse commented:
Regarding the Epic Games ID's, they're usually short but seemingly random words, like "Eider" for HITMAN 3 or "Potoo" for Overcooked! 2. However, I have an idea for obtaining them:

If an issue comes in for a user complaining their Epic game cannot be found, ask them to look inside (or share) their file `~/.local/state/Heroic/logs/runners/legendary.log` (for system-installed) or `~/.var/app/com.heroicgameslauncher.hgl/.local/state/Heroic/logs/runners/legendary.log` (for flatpak). Inside of this log should be a listing of all the games that user owns on Epic, along with their product IDs (listed after `App name:`).

daemon-mouse commented:
I have exhaustively tested this patch with
- (Heroic) Fallout 4,
- (Steam) Fallout New Vegas, and
- (Steam) Fallout 3

This includes MO2 installation, script extended installation, and NXM link handling with or without a running instance of MO2.

However, due to the scope of the changes this patch would benefit from testing by people that own these or other games on different storefronts, if we have the luxury.

In the interest of full disclosure, I couldn't run MO2 from Steam after installing it with (Steam) Fallout 3, nor would NXM links correctly dispatch to (Steam) Fallout New Vegas when MO2 was already running. However, I also have these problems in the official release of 5.2.10, so they aren't regressions. I'll open issues/PRs for these once I can diagnose them well enough.

That said, I'm confident in this patch and excited to share it.

Furglitch commented:
Well this was the last thing I was expecting to see today. Thanks for all the effort you put in.

I'd like to take a closer look before approving, of course, since this is a major overhaul/addition.
I'm currently out of town training for that new job mentioned in that readme disclosure, and right when I get back I'm moving house, so it may not be until next week, but I'll try to get it in ASAP.

daemon-mouse commented:
This patch has now been tested working with:
- (Steam) Oblivion
- (Heroic) Cyberpunk 2077

daemon-mouse committed:
add placeholder URLs for script extenders across supported games

This prevents load_gameinfo.sh from failing when it tries to determine
`game_scriptextender_url` when the store-specific key is missing.

This is a revision of an earlier patch - decided it was better to add
placeholder values.

We could default expansion of
`game_scriptextender_urls` to the empty string, but doing this defeats
the purpose of `set -u` by leaving us unable to detect typos in these
gamesinfo scripts. This way is also more discoverable to the casual
reader how to add a scriptextender to a game.

Furglitch commented:
Everything is looking good! I'm going to merge this into the feature branch, as I'd like to make some slight changes, mostly verbage and variable names, and it'll be easier to test without breaking the master

daemon-mouse commented:
Awesome, thanks! Let me know if you'd like help rebasing this later or if you have any questions!
